File: branches/PrivateBeta/src/PokemonBattle.EffectsP/Items/StateAdded.py
class DeStateBerry(ItemE):
    def __new__(cls, id, state):
        return ItemE.__new__(cls, id)
    def __init__(self, id, state):
        self.State = state
    def e(self, pm):
        if pm.State == self.State:
            pm.DeAbnormalState(True)
    def Attach(self, pm):
        DeStateBerry.e(self, pm)
    def StateAdded(self, pm, by, state):
        DeStateBerry.e(self, pm)
I(DeStateBerry(129, PokemonState.PAR)) #cheri berry
I(DeStateBerry(130, PokemonState.SLP)) #chesto berry
I(DeStateBerry(132, PokemonState.BRN)) #rawst berry
I(DeStateBerry(133, PokemonState.FRZ)) #aspear berry

class PechaBerry(ItemE):
    def e(self, pm):
        if pm.State == PokemonState.PSN or pm.State == PokemonState.BadlyPSN:
            pm.DeAbnormalState(True)
    def Attach(self, pm):
        PechaBerry.e(self, pm)
    def StateAdded(self, pm, by, state):
        if state == AttachedState.PSN:
            PechaBerry.e(self, pm)
I(PechaBerry(131))

class PersimBerry(ItemE):
    def e(self, pm):
        if pm.OnboardPokemon.RemoveCondition('Confuse'):
            pm.ConsumeItem()
            pm.Controller.ReportBuilder.Add(RemoveItem('ItemDeConfuse', pm, 136, 0))
    def Attach(self, pm):
        PerisimBerry.e(self, pm)
    def StateAdded(self, pm, by, state):
        if state == AttachedState.Confusion:
            PerisimBerry.e(self, pm)
I(PersimBerry(136))

class LumBerry(ItemE):
    def e(self, pm):
        if pm.State != PokemonState.Normal:
            pm.DeAbnormalState(True)
    def Attach(self, pm):
        LumBerry.e(self, pm)
    def StateAdded(self, pm, by, state):
        LumBerry.e(self, pm)
I(LumBerry(137))

class DestinyKnot(ItemE):
    def StateAdded(self, pm, by, state):
        if state == AttachedState.Attract:
            by.AddState(pm, AttachedState.Attract, False, 0, 'ItemEnAttract', 57)
I(DestinyKnot(57))

File: branches/PrivateBeta/src/PokemonBattle.Game/Host/Triggers/StateAdded.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace LightStudio.PokemonBattle.Game.Host.Triggers
{
  internal static class StateAdded
  {
    public static void Execute(PokemonProxy pm)
    {
      switch (pm.Item)
      {
        case Is.CHERI_BERRY:
          DeStateBerry(pm, PokemonState.PAR);
          break;
        case Is.CHESTO_BERRY:
          DeStateBerry(pm, PokemonState.SLP);
          break;
        case Is.RAWST_BERRY:
          DeStateBerry(pm, PokemonState.BRN);
          break;
        case Is.ASPEAR_BERRY:
          DeStateBerry(pm, PokemonState.FRZ);
          break;
        case Is.PECHA_BERRY:
          if (pm.State == PokemonState.PSN || pm.State == PokemonState.BadlyPSN) pm.DeAbnormalState(true);
          break;
        case Is.PERSIM_BERRY:
          if (pm.OnboardPokemon.RemoveCondition("Confuse"))
          {
            pm.ConsumeItem();
            pm.RaiseItem("ItemDeConfuse");
          }
          break;
        case Is.LUM_BERRY:
          if (pm.State != PokemonState.Normal) pm.DeAbnormalState(true);
          break;
      }
    }

    private static void DeStateBerry(PokemonProxy pm, PokemonState state)
    {
      if (pm.State == state) pm.DeAbnormalState(true);
    }

    private static bool Act(OnboardPokemon pm, string condition)
    {
      return pm.RemoveCondition(condition);
    }
    private static void Act(PokemonProxy pm)
    {
      var op = pm.OnboardPokemon;
      bool i = Act(op, "Attract");
      bool e = Act(op, "Encore");
      bool ta = Act(op, "Taunt");
      bool to = Act(op, "Torment");
      bool d = Act(op, "Disable");
      if (i || e || ta || to || d)
      {
        pm.ConsumeItem();
        pm.Controller.ReportBuilder.Add(new GameEvents.MentalHerb(pm, i, e, ta, to, d));
      }
    }
  }
}
